modules/apps/fragment/fragment-impl/src/main/java/com/liferay/fragment/internal/processor/PortletRegistryImpl.java

```diff
@@ -188,43 +192,87 @@ public void writePortletPaths(
 		}
 	}
 
-	@Reference(
-		cardinality = ReferenceCardinality.MULTIPLE,
-		policy = ReferencePolicy.DYNAMIC,
-		target = "(com.liferay.fragment.entry.processor.portlet.alias=*)"
-	)
-	protected void setPortlet(
-		javax.portlet.Portlet jxPortlet, Map<String, Object> properties) {
+	@Activate
+	protected void activate(BundleContext bundleContext)
+		throws InvalidSyntaxException {
+
+		String filterString = StringBundler.concat(
+			"(&(com.liferay.fragment.entry.processor.portlet.alias=*)",
+			"(objectClass=", javax.portlet.Portlet.class.getName(), "))");
+
+		_serviceTracker = new ServiceTracker<>(
+			bundleContext, bundleContext.createFilter(filterString),
+			new ServiceTrackerCustomizer
+				<javax.portlet.Portlet, javax.portlet.Portlet>() {
+
+				@Override
+				public javax.portlet.Portlet addingService(
+					ServiceReference<javax.portlet.Portlet> serviceReference) {
+
+					javax.portlet.Portlet jxPortlet = bundleContext.getService(
+						serviceReference);
+
+					String alias = MapUtil.getString(
+						Collections.singletonMap(
+							_keyAlias, serviceReference.getProperty(_keyAlias)),
+						_keyAlias);
+					String portletName = MapUtil.getString(
+						Collections.singletonMap(
+							_keyPortletName,
+							serviceReference.getProperty(_keyPortletName)),
+						_keyPortletName);
+
+					_portletNames.put(alias, portletName);
+
+					bundleContext.registerService(
+						PortletAliasRegistration.class,
+						new PortletAliasRegistration() {
+						},
+						HashMapDictionaryBuilder.<String, Object>put(
+							_keyAlias, alias
+						).build());
+
```

Comment:
@W2jx This is also a bug of the original logic, you register a service here, but you do not unregister when corresponding portlet is undeployed.